### PR TITLE
Update to net 8

### PR DESCRIPTION
**Pull request by Dorus**
#25

.net 8 upgrade. My quick non-scientific test shows no improvements in performance or mem usage (it even got marginally worse in  my tests lol)

Loading my large project:
old
afbeelding
new
afbeelding

As a bonus, i solved:

afbeelding

Currently the project gives [MSB3270](https://learn.microsoft.com/en-us/visualstudio/msbuild/errors/msb3270?view=vs-2022).

This can be solved in 2 ways. Either remove x64 platform target from YAFCparser, or add it to the test or command line project. I've done the latter, considering it was probably added to the parser for good reaons (?).

### CODE REVIEW

**Comment by have-fun-was-taken** [comment]
I seem to not be able to launch YAFC on Windows.
Steps to reproduce:
1. Fetch the PR branch. Check that you got the correct one by `git show head` or any other command.
2. Run `build.bat`.
3. Get `Windows.zip` from the `Build` folder. Unpack it somewhere outside the project folder.
4. Run `YAFC.exe`.
5. YAFC falls silently. The expected behavior is that YAFC launches successfully.

**Comment by have-fun-was-taken** [comment]
Please let me know if you experience the same.

**Comment by Dorus** [comment]
I'm investigating right now. I only tried it from VS 2022, possibly that was naive? Indeed it's not launching here either. I get back on this :)

**Comment by have-fun-was-taken** [comment]
However, I ran the Release build successfully, so it might be the `build.bat` itself.

**Comment by Dorus** [comment]
I tried a few things using a publish profile:

build.bat uses `-p:PublishTrimmed=true` and when i use a publish profile with self-contained+trim unused code, the yafc wont start.
When building without trim and/or framework dependent deployment type (that doesn't support trim), it does work.

Both options have disadvantages i guess, without trim the project doubles in size, with framework dependent it decreases, but i got the feeling it will be framework dependent then (lol).

I'll see if i can find out why trimming fails.

**Comment by Dorus** [comment]
Discoveries so far:

As mentioned above: Disabling trimming works.
When trimming is enabled, to at least start the application, add `<TrimMode>partial</TrimMode>` to the YAFC.csproj
Now json serialization is broken. To [fix this](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/incompatibilities#reflection-based-serializers), we're going to need [source generation](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?pivots=dotnet-8-0#use-source-generation-defaults). First step to get that to work is to remove the line `<LangVersion>8</LangVersion>` both from YAFC.csproj and YAFCui.csproj.

However doing so is not very easy, i got a bit into the source generation, but ran into various problems both with the types that are all needed to be declared and the json settings passed to the serialize functions (It seems to be hard for the analyzers to understand the combination of json settings with TypeInfoResolver)

**Comment by have-fun-was-taken** [comment]
The support of NET Core 6 [goes away](https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core) in November, so no rush. Take your time.

**Comment by Dorus** [comment]
Re #58 @veger 
I'm a bit busy lately, but might give this another swing later. There are 2 ways forwards: Wait till we got a pr for .net with source gen everywhere (i'm half way there in the source-gen branch mentioned above), _or_, disable trimming and make the release 100mb instead of 50m, then work on source gen separate later.

**Comment by have-fun-was-taken** [comment]
On a side note, `dotnet release` writes a veritable wall of warnings about the trimming's ability to break the app, when running `build.bat`.

**Comment by Dorus** [comment]
I'm still biting into this, but for now i cannot figure out how to use code-gen in cases of generic types. The IDE is suggesting to add `DynamicallyAccessedMembers` everywhere, but this is just a red-herring, the said annotation is meant for library authors to communicate their library cannot be used with trimming if said methods are used. Instead we should refactor to get rid of all the reflection calls with the said annotation.

Not to mention i need some features that are currently buggy in .net 8, need to wait for a microsoft service patch.

Today i chased down [line 282 of SerializationMap](https://github.com/have-fun-was-taken/yafc-ce/blob/3e5d79905fb02261d9a2441f700b7925d028eb70/YAFCmodel/Serialization/SerializationMap.cs#L282) as i was expeting to be able to do a simple `new()` here. After chasing these calls down a few methods, adding the `new()` constraint to the generic type, it appears the `ModelObject` we're expecting to receive here doesn't even have a parameterless constructor, yet that's what reflection seems to be calling here. I havent even begun to look at some of the other methods yet, but by now i think it might be more fruitful to just disable trimming at the very least for the models. I'm guessing that was happening in .net 6 and hopefully it's possible in 8 too with some settings.

My work (rebased on latest master) is ongoing in the [source-gen](https://github.com/Dorus/yafc-ce/tree/net8-source-gen) branch.

**Comment by Dorus** [comment]
Added an https://github.com/Dorus/yafc-ce/tree/source-gen .net 6 branch, that would allow part of this pr to be separated if we merge that.

File: Yafc.Model/Yafc.Model.csproj [hunk 1]
```diff
@@ -1,8 +1,8 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
     <PropertyGroup>
-        <RootNamespace>Yafc.Model</RootNamespace>
```
Comment (have-fun-was-taken) [inline Yafc.Model/Yafc.Model.csproj:4]: Was this change automatic? If not, why was this line moved?

File: CommandLineToolExample/CommandLineToolExample.csproj [hunk 2]
```diff
@@ -1,12 +1,12 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
-    <PropertyGroup>
-        <TargetFramework>netcoreapp6.0</TargetFramework>
-        <RootNamespace>YafcCommandLineToolExample</RootNamespace>
-        <LangVersion>8</LangVersion>
```
Comment (have-fun-was-taken) [inline CommandLineToolExample/CommandLineToolExample.csproj:6]: Was this formatting automatic?

Comment (Dorus) [reply CommandLineToolExample/CommandLineToolExample.csproj:6]: Yes, unintentionally but this is what auto-format does for me. 2 spaces per xml property. I intentionally tried not to touch this formatting as much as possible, while still making it consistent among the files themselves. Looks like i failed here since the item group below does have 4 spaces indent.

Do we have a preference here? If so i update the .editorconfig and format all .csproj files that way.

File: Yafc.Model/Yafc.Model.csproj (hunk 1 above)
Comment (Dorus) [reply Yafc.Model/Yafc.Model.csproj:4]: I'm not sure why this line shows as removed, it's not removed.
I did move the TargetFramework around to make it match the same order in al files.

These lines can be removed, but i haven't done so because i want to keep that for if we ever do a all file namespace refactoring. That's outside the scope of this pr. (Namespaces should match folder structure.)

File: CommandLineToolExample/CommandLineToolExample.csproj (hunk 2 above)
Comment (have-fun-was-taken) [reply CommandLineToolExample/CommandLineToolExample.csproj:6]: My preference here would be to keep the XML files as they were. The XML part of editorconfig was not discussed and likely was not applied before because we haven't edited them directly. Feel free to alter editorconfig to reach the previous formatting of XML files.

However, if our XML formatting is unconventional, then we could discuss how to fix it. But I suspect that it is not unconventional because it likley was generated automatically.

Comment (Dorus) [reply CommandLineToolExample/CommandLineToolExample.csproj:6]: Ah, the .editorconfig file already contains 2 spaces for .csproj files

afbeelding

File: Yafc.Model/Yafc.Model.csproj (hunk 1 above)
Comment (have-fun-was-taken) [reply Yafc.Model/Yafc.Model.csproj:4]: The thing that I wanted to clarify is if `<TargetFramework>` being moved above `<RootNamespace>` was done manually or automatically. Before the change the TargetFramework was below, so I am curious if it was moved automatically.

File: CommandLineToolExample/CommandLineToolExample.csproj (hunk 2 above)
Comment (Dorus) [reply CommandLineToolExample/CommandLineToolExample.csproj:6]: I ran some tests:

Visual studio scaffolding of .csproj files prefers 2 spaces independent of .editorconfig

Visual studio property editor prefers to respect existing space usage. So if the file has 4 spaces already for some xml nodes, new ones will be added at the same spacing.

Similar to RootNamespace. Visual studio prefers to keep it where it is, but will move it below TargetFramework if it can add it.

What i think is going on here is that these xml files are created by different versions of visual studio over the years, and the scaffolding might have contained different defaults before. It's also possible some manual reformatting has happened here and there.

The formatting itself doesn't harm anything, except readability in a file nobody is supposed to read. And it makes pull request change sets more annoying to review. So mostly my changes moved the files closer to what they would have been by default if created today. The main question remains if we want to take those changes with this Pr, or defer them to some other time?

File: Yafc.Model/Yafc.Model.csproj (hunk 1 above)
Comment (Dorus) [reply Yafc.Model/Yafc.Model.csproj:4]: Yes, as i said:

> I did move the TargetFramework around to make it match the same order in al files.

If reordering is inappropriate for this pr i can revert it, but i found keeping the order consistent between files easier to work with.

Comment (have-fun-was-taken) [reply Yafc.Model/Yafc.Model.csproj:4]: If the order is the same in the other files, then it's fine.

File: CommandLineToolExample/CommandLineToolExample.csproj (hunk 2 above)
Comment (have-fun-was-taken) [reply CommandLineToolExample/CommandLineToolExample.csproj:6]: Thank you for the clarificaiton! The XML reformat can be done later.

**Comment by have-fun-was-taken** [comment]
Surprisingly, the build size decreases with this PR:
Old size was 62MB, 27MB when compressed.
New size is 39MB, 16MB when compressed.

**Review by have-fun-was-taken — APPROVED** [review]
Thank you for the contribution!

In the future, please keep the formatting consistent with the rest of the file. That allows the diff to show the change in the logic instead of showing the whole reformatted part.

I would be thankful for an XML-reformat-PR after that because with this PR, the formatting is inconsistent.

**Comment by have-fun-was-taken** [comment]
@Dorus please merge the PR with an informative commit message when you're ready.

**Comment by have-fun-was-taken** [comment]
> In the future, please keep the formatting consistent with the rest of the file. That allows the diff to show the change in the logic instead of showing the whole reformatted part.

Or rather just keep the reformatting commit separate from the commit with the changes.

**Comment by Dorus** [comment]
> > In the future, please keep the formatting consistent with the rest of the file. That allows the diff to show the change in the logic instead of showing the whole reformatted part.
> 
> Or rather just keep the reformatting commit separate from the commit with the changes.

That's a very fair suggestion. Will do so from now on.
